File: ios/YiLuAn/Core/Models/FamilyMember.swift
import Foundation

/// [F-05] 代他人下单：家人档案（后端 GET /api/v1/users/me/family-members）
struct FamilyMember: Codable, Identifiable, Hashable {
    let id: String
    let userId: String
    let name: String
    let relation: String?
    let phone: String?
    let gender: String?
    let age: Int?
    let medicalNotes: String?
    let createdAt: Date?
    let updatedAt: Date?

    var relationLabel: String { FamilyRelation.label(for: relation) }
}

struct FamilyMemberListResponse: Decodable {
    let items: [FamilyMember]
    let total: Int
}

/// Request body for POST/PATCH /users/me/family-members.
/// 全部 optional 以适配 PATCH 部分更新；POST 时 name 必传。
struct FamilyMemberRequest: Encodable {
    var name: String?
    var relation: String?
    var phone: String?
    var gender: String?
    var age: Int?
    var medicalNotes: String?
}

/// Mirrors backend FamilyRelation enum.
enum FamilyRelation {
    /// relation code 列表(显示文案经 loc 字典, 避免 Model 层硬编码中文 C 盲区)
    static let allValues: [String] = [
        "parent", "spouse", "child", "sibling",
        "grandparent", "relative", "friend", "other",
    ]

    /// (value, 本地化 label) 对 —— label 运行时走字典(shared 单例, Model 无 @EnvironmentObject scope)
    static var allCases: [(value: String, label: String)] {
        allValues.map { ($0, label(for: $0)) }
    }

    static func label(for value: String?) -> String {
        let loc = LocalizationManager.shared
        guard let v = value else { return loc.t("relation.other") }
        return loc.t("relation.\(v)")
    }
}

enum FamilyGender: String, CaseIterable {
    case unknown
    case male
    case female

    var label: String {
        let loc = LocalizationManager.shared
        switch self {
        case .unknown: return loc.t("familyMembers.genderUnknown")
        case .male: return loc.t("familyMembers.genderMale")
        case .female: return loc.t("familyMembers.genderFemale")
        }
    }
}

File: backend/app/models/family_member.py
"""F-05: family members (代他人下单).

A *family member* is a non-account person that a user can place orders on
behalf of (typically an elderly parent). The booking user remains the
order owner (patient_id == booking user), but the order is denormalized
with the family member's name / relation / phone so the companion knows
who they are actually accompanying.

Hard rules (D-056):
* `user_id` is indexed; one user may have many members.
* `deleted_at` for soft-delete — historical orders keep their reference.
* No id-card / no shared payment / no inter-member transfer in MVP.
"""

from __future__ import annotations

import enum
import uuid
from datetime import datetime, timezone

from sqlalchemy import DateTime, Enum, Integer, String, Text, Uuid
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class FamilyRelation(str, enum.Enum):
    self_ = "self"  # rarely used; we usually skip the row entirely
    parent = "parent"
    spouse = "spouse"
    child = "child"
    sibling = "sibling"
    grandparent = "grandparent"
    relative = "relative"
    friend = "friend"
    other = "other"


class FamilyGender(str, enum.Enum):
    unknown = "unknown"
    male = "male"
    female = "female"


class FamilyMember(Base):
    __tablename__ = "family_members"

    id: Mapped[uuid.UUID] = mapped_column(
        Uuid(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        Uuid(as_uuid=True), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(50), nullable=False)
    relation: Mapped[FamilyRelation] = mapped_column(
        Enum(FamilyRelation),
        nullable=False,
        default=FamilyRelation.other,
        server_default=FamilyRelation.other.value,
    )
    phone: Mapped[str | None] = mapped_column(String(20), nullable=True)
    gender: Mapped[FamilyGender] = mapped_column(
        Enum(FamilyGender),
        nullable=False,
        default=FamilyGender.unknown,
        server_default=FamilyGender.unknown.value,
    )
    age: Mapped[int | None] = mapped_column(Integer, nullable=True)
    medical_notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    deleted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, index=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
